### DISCUSSION TITLE
How to set the font-size of text inside a Button?

### DISCUSSION THREAD

**Discussion by fancellu** [slint-ui/slint#4325] [Q&A]
I ask as the default font size seems tiny on my 4k screen, and I can't adjust it as I can for a simple Text

e.g.

```
import { Button, VerticalBox } from "std-widgets.slint";

export component AppWindow inherits Window {
    in-out property<int> counter: 42;
    callback request-increase-value();
    VerticalBox {
           
        Text {
            text: "Counter: \{root.counter}";
            font-size: 50px;
        }
        Button {
            text: "Increase value!!!";
            
            clicked => {
                root.request-increase-value();
            }
        }
    }
}
```

Trust me, it is much smaller on my 4k screen. 

![image](https://github.com/slint-ui/slint/assets/4273247/43128c65-3ff6-4dd7-bd32-0e2e2839d2a2)

**Comment by bjorn** [comment]
You may want to try setting `default-font-size`: https://slint.dev/releases/1.3.2/docs/slint/src/language/concepts/fonts

**Reply by sealor** [reply]
Unfortunately, that does not work.

```rust
slint::slint! {
    import { Button, VerticalBox } from "std-widgets.slint";
    export component BrowserChooser inherits Window {
        default-font-size: 30px;
        VerticalBox {
            Button {
                text: "hello world";
            }
            Text {
                text: "hello world";
            }
        }
    }
}

fn main() {
    BrowserChooser::new().unwrap().run().unwrap();
}
```
```toml
[package]
name = "font-size-test"
version = "0.1.0"
edition = "2021"

[dependencies]
slint = "1.4.1"
```
Result:
![Bildschirmfoto_2024-02-03_12-41-22](https://github.com/slint-ui/slint/assets/6491612/a70ad203-799b-40b6-b802-4cb0cf1758f2)

**Reply by ogoffart** [reply]
I think the Qt style don't respect the default window style.

In practice, you can't change the font size of a native button because the font size should be the one which is good for buttons.  If you want non-native style, you need to re-implement your components.

If the font is too small for native component, this can mean that the detection of the scale factor is wrong.  What is your platform? Is it wayland or x11?  Do you have a scale factor set?  Can you also try with the winit backend?

**Reply by sealor** [reply]
Thank you for your reply.
I'm using Linux (EndeavourOS) with xfce4 (X11). I didn't set a scale factor. I only used the code which I wrote previously.

When using the `winit` backend the font size is higher by default but the `default-font-size` attribute is still not recognized by the button text.

`SLINT_BACKEND=winit cargo run`

![Bildschirmfoto_2024-02-04_19-04-11](https://github.com/slint-ui/slint/assets/6491612/317fd99f-4c17-4ce5-9e94-ef1f70a6c994)

**Reply by tronical** [reply]
Even with the winit backend this is still rendered using the Qt style. We could make it respect default-font-size if we use `get_font()` like we do for `draw_text()` and pass the `QFont` to `QStyleOption`.

@sealor If you [set the widget style](https://slint.dev/releases/1.4.1/docs/slint/src/advanced/style#selecting-a-widget-style-with-rust) to say "Fluent", then it `default-font-size` should work.

@ogoffart would you say that it's a bug that the QStyle based style doesn't respect `default-font-*` or would you say that it's by design?

**Reply by ogoffart** [reply]
Yes, this is probably a bug that the qt style don't respect the style.
But even then, you don't want to encourage people to set it. 
Because
 1. That's not the way to do it so a deskop app looks good on a 4k monitor. Normally, it should use the proper scale factor such that the font text looks the same as every apps. And it would look terribly wrong when deployed on a non-4k monitor.
 2. that doesn't work in the live preview when previewing components separately.

**Reply by tronical** [reply]
Related issue to track support for `default-font-*` with the Qt style: https://github.com/slint-ui/slint/issues/1022

**Comment by sealor** [comment]
> Even with the winit backend this is still rendered using the Qt style. We could make it respect default-font-size if we use `get_font()` like we do for `draw_text()` and pass the `QFont` to `QStyleOption`.
> 
> @sealor If you [set the widget style](https://slint.dev/releases/1.4.1/docs/slint/src/advanced/style#selecting-a-widget-style-with-rust) to say "Fluent", then it `default-font-size` should work.
> 
> @ogoffart would you say that it's a bug that the QStyle based style doesn't respect `default-font-*` or would you say that it's by design?

I tried the following:
```
SLINT_STYLE=fluent cargo run
```
Result:
![Bildschirmfoto_2024-02-06_21-46-54](https://github.com/slint-ui/slint/assets/6491612/7d998979-cd31-4969-8404-7d05627f533e)
